File: vscode-extension/src/diagramDocument.ts
import {
  applyEdits,
  modify,
  parse,
  type FormattingOptions,
  type ParseError,
} from "jsonc-parser"
import type { UMLDiagramType, UMLModel } from "@tumaet/apollon"

/** The `UMLModel` wire-format line this extension reads and writes. */
const MODEL_SCHEMA_VERSION = "4.0.0"

/** Older `.apollon` files nest the model under this key alongside app metadata. */
const MODEL_KEY = "model"

export class DiagramParseError extends Error {}

const isRecord = (v: unknown): v is Record<string, unknown> =>
  typeof v === "object" && v !== null && !Array.isArray(v)

const isWrapped = (root: unknown): root is Record<string, unknown> =>
  isRecord(root) && MODEL_KEY in root

/**
 * Where the model lives in a given document. New files are a bare `UMLModel`;
 * files written by earlier versions wrap it under `model` next to now-unused
 * app metadata. Editing in place at this path keeps the wrapper's other keys and
 * their formatting untouched, so a diagram edit never rewrites the whole file.
 */
const modelPath = (root: unknown): (string | number)[] =>
  isWrapped(root) ? [MODEL_KEY] : []

/** Reads the model out of a `.apollon` document, wrapped or bare. */
export function parseModel(text: string): UMLModel {
  const errors: ParseError[] = []
  const root: unknown = parse(text, errors, { allowTrailingComma: true })
  if (errors.length > 0 || root === undefined) {
    throw new DiagramParseError("not valid JSON")
  }
  const candidate = isWrapped(root) ? root[MODEL_KEY] : root
  if (!isRecord(candidate) || typeof candidate.type !== "string") {
    throw new DiagramParseError("no UML model with a `type` field")
  }
  // Only the discriminator is checked here. Structural validation belongs to the
  // library, which rejects a malformed model on mount with a better message than
  // anything this extension could reconstruct from a JSON shape.
  return candidate as unknown as UMLModel
}

/**
 * What a document holds, as a value rather than an exception. An empty file is
 * not a failure: it is a diagram waiting to be chosen, and the canvas says so.
 */
export type DocumentState =
  | { kind: "empty" }
  | { kind: "model"; model: UMLModel }
  | { kind: "invalid"; reason: string }

export function readDocument(text: string): DocumentState {
  if (text.trim() === "") {
    return { kind: "empty" }
  }
  try {
    return { kind: "model", model: parseModel(text) }
  } catch (error) {
    return {
      kind: "invalid",
      reason:
        error instanceof DiagramParseError ? error.message : String(error),
    }
  }
}

/** A replacement of `[offset, offset + length)` — `jsonc-parser`'s edit shape. */
export interface TextRangeEdit {
  offset: number
  length: number
  content: string
}

/**
 * The edits that write `model` into `text`, touching only the model's own range.
 * Sibling keys, key order and the file's existing indentation survive, and the
 * range is minimal — so a text editor open on the same document keeps its cursor,
 * and undo granularity matches what actually changed.
 */
export function modelEdits(text: string, model: UMLModel): TextRangeEdit[] {
  const root: unknown = parse(text, [], { allowTrailingComma: true })
  return modify(text, modelPath(root), model, {
    formattingOptions: detectFormatting(text),
  })
}

export const applyTextRangeEdits = (
  text: string,
  edits: TextRangeEdit[]
): string => applyEdits(text, edits)

/** The model a brand-new diagram starts from. */
export function scaffoldModel(type: UMLDiagramType, title: string): UMLModel {
  return {
    version: MODEL_SCHEMA_VERSION,
    id: crypto.randomUUID(),
    title,
    type,
    nodes: [],
    edges: [],
    assessments: {},
  }
}

/** {@link scaffoldModel} as the bytes of a new `.apollon` file. */
export function scaffold(type: UMLDiagramType, title: string): string {
  return `${JSON.stringify(scaffoldModel(type, title), null, 2)}\n`
}

/** Match the document's own indentation and line endings rather than imposing ours. */
export function detectFormatting(text: string): FormattingOptions {
  const eol = text.includes("\r\n") ? "\r\n" : "\n"
  const indent = /\n([ \t]+)\S/.exec(text)?.[1]
  if (indent?.startsWith("\t")) {
    return { tabSize: 2, insertSpaces: false, eol }
  }
  return { tabSize: indent?.length ?? 2, insertSpaces: true, eol }
}

/** Sibling image path for a diagram — `diagram.apollon` → `diagram.svg`. */
export function exportTargetPath(path: string, extension: string): string {
  return `${path.replace(/\.[^./\\]*$/, "")}.${extension}`
}

File: jetbrains-plugin/src/main/kotlin/de/tum/cit/aet/apollon/document/DiagramDocument.kt
package de.tum.cit.aet.apollon.document

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import java.util.UUID

/** The `UMLModel` wire-format line this plugin reads and writes. */
private const val MODEL_SCHEMA_VERSION = "4.0.0"

/** Older `.apollon` files nest the model under this key alongside app metadata. */
private const val MODEL_KEY = "model"

class DiagramParseError(message: String) : Exception(message)

/**
 * What a document holds, as a value rather than an exception. An empty file is
 * not a failure: it is a diagram waiting to be chosen, and the canvas says so.
 */
sealed interface DocumentState {
    data object Empty : DocumentState

    data class Model(val model: JsonObject) : DocumentState

    data class Invalid(val reason: String) : DocumentState
}

private fun isWrapped(root: JsonElement): Boolean = root is JsonObject && MODEL_KEY in root

/** Reads the model out of a `.apollon` document's text, wrapped or bare. */
fun parseModel(text: String): JsonObject {
    val root =
        try {
            Json.parseToJsonElement(text)
        } catch (e: Exception) {
            throw DiagramParseError("not valid JSON")
        }
    val candidate = if (isWrapped(root)) root.jsonObject[MODEL_KEY] else root
    val type = (candidate as? JsonObject)?.get("type")
    if (candidate !is JsonObject || type !is JsonPrimitive || !type.isString) {
        throw DiagramParseError("no UML model with a `type` field")
    }
    return candidate
}

fun readDocument(text: String): DocumentState {
    if (text.isBlank()) {
        return DocumentState.Empty
    }
    return try {
        DocumentState.Model(parseModel(text))
    } catch (e: DiagramParseError) {
        DocumentState.Invalid(e.message ?: "invalid")
    }
}

data class Formatting(val indent: String, val eol: String)

/** Match the document's own indentation and line endings rather than imposing ours. */
fun detectFormatting(text: String): Formatting {
    val eol = if (text.contains("\r\n")) "\r\n" else "\n"
    val indent = Regex("\n([ \t]+)\\S").find(text)?.groupValues?.get(1)
    return Formatting(indent ?: "  ", eol)
}

/**
 * The bytes to write for `model` given the document's current text.
 *
 * The VS Code extension edits only the model's own byte range via
 * `jsonc-parser`, preserving sibling keys, key order and untouched formatting
 * exactly. There is no Kotlin equivalent of that library, so this rewrites
 * the whole document instead (Option A from the implementation plan): sibling
 * keys of a legacy wrapped file survive with their *values* and relative
 * order intact (rebuilt via [JsonObject], which preserves insertion order),
 * but comments, trailing commas and this-key's-exact-original-formatting do
 * not. A `.apollon` file this plugin has written is always the bare shape;
 * only a file it merely opens once (written by an older tool) can still be
 * wrapped, and normalizes to bare the first time this plugin edits it.
 */
fun writeDocumentText(
    existingText: String,
    model: JsonObject,
): String {
    val formatting = detectFormatting(existingText)
    val root =
        if (existingText.isBlank()) {
            null
        } else {
            runCatching { Json.parseToJsonElement(existingText) }.getOrNull()
        }
    val output: JsonObject =
        if (root is JsonObject && MODEL_KEY in root) {
            JsonObject(root.toMutableMap().apply { put(MODEL_KEY, model) })
        } else {
            model
        }
    val printer = Json { prettyPrint = true; prettyPrintIndent = formatting.indent }
    val body = printer.encodeToString(JsonObject.serializer(), output)
    val withEol = if (formatting.eol == "\r\n") body.replace("\n", "\r\n") else body
    return "$withEol${formatting.eol}"
}

/** The model a brand-new diagram starts from. */
fun scaffoldModel(
    diagramType: String,
    title: String,
): JsonObject =
    buildJsonObject {
        put("version", MODEL_SCHEMA_VERSION)
        put("id", UUID.randomUUID().toString())
        put("title", title)
        put("type", diagramType)
        put("nodes", buildJsonArray {})
        put("edges", buildJsonArray {})
        put("assessments", buildJsonObject {})
    }

/** [scaffoldModel] as the bytes of a new `.apollon` file. */
fun scaffoldDocument(
    diagramType: String,
    title: String,
): String = writeDocumentText("", scaffoldModel(diagramType, title))

/** Sibling image path for a diagram — `diagram.apollon` -> `diagram.svg`. */
fun exportTargetPath(
    path: String,
    extension: String,
): String = Regex("\\.[^./\\\\]*$").replace(path, "") + ".$extension"

/** The diagram a path names — `.../orders.apollon` -> `orders`. */
fun diagramTitle(path: String): String =
    path.substringAfterLast('/').replace(Regex("\\.apollon$", RegexOption.IGNORE_CASE), "")
